examples/anomaly_detection/deep_sad/model_def.py

```diff
@@ -0,0 +1,267 @@
+            )
+
+            loss = torch.mean(rec_loss)
+            epoch_loss += loss.item()
+            n_batches += 1
+
+        _, labels, scores = zip(*idx_label_score)
+        labels = np.array(labels)
+        scores = np.array(scores)
+        test_auc = roc_auc_score(labels, scores)
+
+        # Log results
+        validation_loss = epoch_loss / n_batches
+        print("Test Loss: {:.6f}".format(validation_loss))
+        print("Test AUC: {:.2f}%".format(100.0 * test_auc))
+        print("Finished testing autoencoder.")
+
+        return {
+            "validation_loss": validation_loss,
+            "test_auc": test_auc,
+        }
+
+
+def get_filtered_ae_checkpoint_state_dict():
+    # Extract encoder model weights so they can be loaded back.
+    checkpoint_path = Path(__file__).resolve().parent / "ae_state_dict.pth"
+    checkpoint_data = torch.load(checkpoint_path)
+    state_dict = checkpoint_data["models_state_dict"][0]
+    state_dict_pure = OrderedDict()
+    for k, v in state_dict.items():
+        if k.startswith("encoder."):
+            state_dict_pure[k.replace("encoder.", "")] = v
+
+    return state_dict_pure
+
+
+class DeepSADMainTrial(DeepSADAutoEncoderTrial):
+    def __init__(self, context: PyTorchTrialContext):
+        self.context = context
+
+        self.model = self.context.wrap_model(MNIST_LeNet())
+        self.model.load_state_dict(get_filtered_ae_checkpoint_state_dict())
+
+        self.optimizer = self.context.wrap_optimizer(
+            torch.optim.Adam(
+                self.model.parameters(),
+                lr=self.context.get_hparam("learning_rate"),
+                weight_decay=self.context.get_hparam("weight_decay"),
+            )
+        )
+
+        self.lr_scheduler = self.context.wrap_lr_scheduler(
+            torch.optim.lr_scheduler.MultiStepLR(
+                self.optimizer, milestones=(50, 100), gamma=0.1
+            ),
+            step_mode=LRScheduler.StepMode.STEP_EVERY_EPOCH,
+        )
+
+        self.download_directory = f"/tmp/data-rank{self.context.distributed.get_rank()}"
+        self.dataset: Optional[TorchvisionDataset] = None
+
+        self.eps = 1e-6
+        self.eta = self.context.get_hparam("eta")
+
+    def build_training_data_loader(self) -> pytorch.DataLoader:
+        res = super().build_training_data_loader()
+        self.c = self.init_center_c(res.get_data_loader())
+        return res
+
+    def train_batch(
+        self, batch: pytorch.TorchData, epoch_idx: int, batch_idx: int
+    ) -> Union[torch.Tensor, Dict[str, Any]]:
+        inputs, _, semi_targets, _ = batch
+        self.optimizer.zero_grad()
+
+        outputs = self.model(inputs)
+        dist = torch.sum((outputs - self.c) ** 2, dim=1)
+        losses = torch.where(
+            semi_targets == 0,
+            dist,
+            self.eta * ((dist + self.eps) ** semi_targets.float()),
+        )
+        loss = torch.mean(losses)
+        loss.backward()
+        self.optimizer.step()
```

Comment:
Is there a reason not to use the contex.backwards here?

Reply:
good catch. it's one of the ambiguous porting points. under what circumstances would someone still use `loss.backward` and raw `optimizer.step()` instead of our context way? can/should we warn the user somehow?